YubiKit/YubiKit/Helpers/YKFTLVRecord.m

```diff
@@ -0,0 +1,155 @@
+            length = (length << 8) | (bytes[offset++] & 0xff);
+        }
+    }
+    // data
+    *bytesRead = offset + length;
+    if (checkMatchingLength && data.length != offset + length) { return nil; }
+    if (data.length < offset + length) { return nil; }
+    return [[YKFTLVRecord alloc] initWithTag:tag value:[data subdataWithRange:NSMakeRange(offset, length)]];
+}
+
+- (NSData *)data {
+
+    NSMutableData * result = [NSMutableData new];
+    
+    // tag
+    YKFTLVTag tag = CFSwapInt32HostToBig(self.tag);
+    char* tagBytes = (char*) &tag;
+    NSData *tagData = [NSData ykf_encodeBytesBigSkippingZeros:tagBytes length:sizeof(YKFTLVTag)];
+    [result appendData:tagData];
+    
+    // length
+    NSUInteger hostLength = self.value.length;
+    if (hostLength < 0x80) {
+        [result appendBytes:&hostLength length:1];
+    } else {
+        NSUInteger length = NSSwapHostLongToBig(hostLength);
+        char* lengthBytes = (char*)&length;
+        NSData *lengthData = [NSData ykf_encodeBytesBigSkippingZeros:lengthBytes length:sizeof(length)];
+        Byte lengthHeader = 0x80 | lengthData.length;
+        [result appendBytes:&lengthHeader length:1];
+        [result appendData:lengthData];
+    }
+    
+    // data
+    [result appendData:self.value];
+
+    return result;
+}
+
+- (instancetype _Nonnull )initWithTag:(YKFTLVTag)tag value:(NSData *_Nonnull)value {
+    self = [super init];
+    if (self) {
+        self.tag = tag;
+        self.value = value;
+    }
+    return self;
+}
+
+- (instancetype _Nonnull )initWithTag:(YKFTLVTag)tag records:(NSArray<YKFTLVRecord *> *_Nonnull)records {
+    NSMutableData *data = [NSMutableData new];
+    for (YKFTLVRecord * record in records) {
+        [data appendData:record.data];
+    }
+    return [[YKFTLVRecord alloc] initWithTag:tag value:data];
+}
+
++ (nullable instancetype)recordFromData:(NSData *_Nullable)data {
+    int bytesRead = 0;
+    return [self recordFromData:data checkMatchingLength:true bytesRead:&bytesRead];
+}
+
++ (nullable NSArray<YKFTLVRecord *> *)sequenceOfRecordsFromData:(NSData *_Nullable)data {
+    
+    NSMutableArray<YKFTLVRecord *> *records = [[NSMutableArray<YKFTLVRecord *> alloc] init];
+    
+    int location = 0;
+    bool keepScanning = true;
+    while (keepScanning) {
+        data = [data subdataWithRange:NSMakeRange(location, data.length - location)];
+        
+        int bytesRead = 0;
+        YKFTLVRecord *record = [YKFTLVRecord recordFromData:data checkMatchingLength:NO bytesRead:&bytesRead];
+        if (record) {
+            [records addObject:record];
+            location = bytesRead;
+            if (location >= data.length) {
+                keepScanning = NO;
+            }
+        } else {
+            records = nil;
+            break;
+        }
+    }
+    return records;
+}
+
+@end
+
+
+@implementation NSData (NSData_ByteEncoding)
+    
++ (NSData *)ykf_encodeBytesBigSkippingZeros:(char*)bytes length:(int)length {
```

Comment:
Encoding the integer as bytes and byte order is handled outside of this function, so I would rename it to just `ykf_stripLeadingZeros` or something like that. That, or change the argument to the function to be the integer itself rather than the (already encoded) byte array.

Reply:
Yea, I was debating that with myself as well. I'm changing it to `ykf_dataWithBytesStripLeadingZeros` to better align with the methods already implemented on `NSData`.